### DISCUSSION TITLE
Looking at the broken falling animations

### DISCUSSION THREAD

**Discussion by lewri** [CorsixTH/CorsixTH#1846] [General Diagnostics]
Noted in the code of patient.lua is that falling during an earthquake is broken, though the proper explanation as to why is in world:
https://github.com/CorsixTH/CorsixTH/blob/faa908ec368cc6b6b00fc38413265147b49ed93f/CorsixTH/Lua/world.lua#L660-L670

The code for use for falls looks like below:
https://github.com/CorsixTH/CorsixTH/blob/faa908ec368cc6b6b00fc38413265147b49ed93f/CorsixTH/Lua/entities/humanoids/patient.lua#L340-L365

https://github.com/CorsixTH/CorsixTH/blob/faa908ec368cc6b6b00fc38413265147b49ed93f/CorsixTH/Lua/humanoid_actions/falling.lua#L26-L33
*Note that Line 28 should be ``self:setMustHappen(true)``*

Yet pointed out the "error" of nil action queues seems to happen when trying to get back up:

https://github.com/CorsixTH/CorsixTH/blob/faa908ec368cc6b6b00fc38413265147b49ed93f/CorsixTH/Lua/humanoid_actions/get_up.lua#L26-L32

```
Empty action queue!
Last action: get_up
stack traceback:
        Lua\entities\humanoid.lua:500: in upvalue 'Humanoid_startAction'
        Lua\entities\humanoid.lua:634: in method 'finishAction'
        Lua\humanoid_actions\get_up.lua:31: in local 'timer_function'
        Lua\entity.lua:216: in method 'tick'
        Lua\world.lua:1070: in method 'onTick'
        Lua\app.lua:1140: in function <Lua\app.lua:1137>
        (...tail calls...)
        Lua\app.lua:1040: in function <Lua\app.lua:1035>
```

My hunch is the failure is due to poor preservation of existing action queue items. ``Patient:interruptAndRequeueAction`` seems to be the function meant to handle this, but it will currently resolve to thinking the patient was looking for a room/door in its current state.

https://github.com/CorsixTH/CorsixTH/blob/faa908ec368cc6b6b00fc38413265147b49ed93f/CorsixTH/Lua/entities/humanoids/patient.lua#L1102-L1147

Safe to say, turning the function back off for now!

**Comment by mugmuggy** [comment]
Yeah I have a fix for falling in progress, it still needs finesse around the action, can they fall multiple times etc.

What I noted and it might have been on an issue or an irc discussion, was the interruptAndRequeueAction (I introduced this function on a refactor but the issue is fundamental) will on occasion requeue an action that should have been interrupted. As the on_interrupt handler is removed from the action, then called, subsequent interruptions though setNextAction (on ground/getup should just be queued actions) for example will call finishAction - same issue we have with the queue interruptions and drinks machines use animation jumping. The WalkAction recreates the on_interrupt if its not to be interrupted.

Ignoring this, the falling was failing as it could interrupt any falling action (broken 'must happen's) and indeed each of these are 'interrupted' due to the setNextAction calls. Also female patient on ground animation was wrong, was the get up animation.

I'll have to try and get back into the swing of things and get these changes up.

**Reply by lewri** [reply]
Nice! So the breakage of things was happening from interrupting an already happening fall? (Though potentially more).
[Also we miss you on IRC]

**Comment by lewri** [comment]
RE your other comment about multiple times in a quake. I think in the original they would only fall once per quake... Or the falling chance was so low they only ever fell once

**Reply by mugmuggy** [reply]
Yep that also matches what I recall the behaivour is, but the falling chance is the interesting one, I don't know if the falling chance is low or just something influences it enough not to allow it to happen when falling is initiated in TH. Corsix has everyone falling over. I'll have to take some observations again.
